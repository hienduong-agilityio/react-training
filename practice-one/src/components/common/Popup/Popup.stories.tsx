import Popup from './Popup';

export default {
  title: 'Popup',
  component: Popup,
};

export const Default = () => (
  <Popup isOpen={true} onClose={() => {}}>
    <div>
      <h2>Popup Content</h2>
      <p>This is a sample content for the Popup component.</p>
    </div>
  </Popup>
);

export const Closed = () => (
  <Popup isOpen={false} onClose={() => {}}>
    <div>
      <h2>Popup Content</h2>
      <p>This Popup is closed.</p>
    </div>
  </Popup>
);
