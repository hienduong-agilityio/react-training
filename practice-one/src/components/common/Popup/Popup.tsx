import classNames from 'classnames';
import React from 'react';
import styles from './index.module.css';
import Button from '../Button/Button';

export interface PopupProps {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
  className?: string;
}

const Popup = ({ isOpen, onClose, children, className }: PopupProps) => {
  const overlayClasses = classNames(styles.overlay, className);
  if (!isOpen) return null;

  return (
    <div className={overlayClasses}>
      <div className={styles.popup}>
        <Button customClasses={styles.closeButton} onClick={onClose}>
          Close
        </Button>
        <div className={styles.content}>{children}</div>
      </div>
    </div>
  );
};

export default Popup;
