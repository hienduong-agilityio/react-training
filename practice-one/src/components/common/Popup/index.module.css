

.container {
  z-index: 30;
}

.popup {
  position: absolute;
  display: flex;
  top: 10;
  left: -5;
  z-index: 200;
  background-color: var(--white-color);
  padding: 20px;
  border-radius: 5px;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.3);
  width: 400px;
  max-width: 90%;
}

.closeButton {
  background-color: var(--primary-color);
  color: var(--white-color);
  border: none;
  padding: 8px 16px;
  border-radius: 4px;
  cursor: pointer;
  margin-bottom: 10px;
}

.closeButton:hover {
  background-color: var(--active-color);
}
