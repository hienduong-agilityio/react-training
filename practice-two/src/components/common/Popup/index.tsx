interface IPopupProps {
  isOpen: boolean;
  children?: React.ReactNode | string;
  onClosePopup?: () => void;
}

/**
 * Popup component to display content in a modal-like dialog
 * @param isOpen - Indicates whether the popup is open or closed
 * @param children - Content to be displayed inside the popup
 * @param onClosePopup - Callback function to handle popup close event
 *
 * @returns {JSX.Element} Popup component
 */

export default function Popup({
  isOpen = false,
  children,
  onClosePopup = () => {}
}: Readonly<IPopupProps>): JSX.Element {
  return (
    <div
      onClick={onClosePopup}
      className={`fixed inset-0 flex justify-center items-center transition-colors ${isOpen ? 'block bg-black/20' : 'hidden'}`}
    >
      {/* popup */}
      <div
        onClick={(e) => e.stopPropagation()}
        className={`bg-white rounded-xl p-6 z-50 ${isOpen ? 'scale-100 opacity-100 z-50' : 'scale-125 opacity-0 z-50'}`}
      >
        {children}
      </div>
    </div>
  );
}
