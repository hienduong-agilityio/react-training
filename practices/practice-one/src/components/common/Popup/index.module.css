.container {
  z-index: 30;
}
.overlay {
  position: fixed;
  inset: 0;
  background-color: rgba(0, 0, 0, 0.7);
  backdrop-filter: blur(8px);
  display: flex;
  align-items: center;
  justify-content: center;
}

.popupDefault {
  position: relative;
  display: flex;
  top: 10px;
  left: -5px;
  z-index: 200;
  background-color: var(--white-color);
  padding: 20px;
  border-radius: 5px;
  height: 90vh;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.3);
  width: 400px;
  max-width: 100%;
}

.closeButton {
  background-color: var(--icon-color);
  color: var(--white-color);
  border: none;
  padding: 8px 16px;
  border-radius: 4px;
  cursor: pointer;
  margin-bottom: 10px;
}

.closeButtonContent {
  position: absolute;
  top: 10px;
  right: 15px;
}
