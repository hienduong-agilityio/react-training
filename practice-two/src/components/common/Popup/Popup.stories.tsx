import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react';

// Components
import Popup from './index';
import Button from '../Button';

// Meta information for Storybook
export default {
  title: 'Components/Popup',
  component: Popup,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs']
} as Meta;

// Define args interface
interface ModalProps {
  closeButtonContent?: string;
}

// Define Template
const Template: Story<ModalProps> = (args) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <Button onClick={() => setIsOpen(true)}>Show Popup</Button>
      <Popup
        isOpen={isOpen}
        onClosePopup={() => setIsOpen(false)}
        closeButtonContent={args.closeButtonContent}
        closeButton
      >
        <div className="text-center w-56">
          <div className="mx-auto my-4 w-48">
            <h3 className="text-lg font-black text-gray-800">Confirm Delete</h3>
            <p className="text-sm text-gray-500">Are you sure you want to delete this item?</p>
          </div>
          <div className="flex gap-4">
            <Button customClasses="bg-danger" onClick={() => setIsOpen(false)}>
              confirm
            </Button>
            <Button customClasses="bg-primary">Cancel</Button>
          </div>
        </div>
      </Popup>
    </div>
  );
};

export const Default = Template.bind({});
Default.args = {
  closeButtonContent: 'Close'
};
