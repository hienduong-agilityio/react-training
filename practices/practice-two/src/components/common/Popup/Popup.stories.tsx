import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react';

// Components
import Popup from './index';
import Button from '../Button';

// Meta information for Storybook
export default {
  title: 'Components/Popup',
  component: Popup,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs']
} as Meta;

// Define args interface
interface ModalProps {
  closeButtonContent?: string;
}

// Define Template
const Template: Story<ModalProps> = () => {
  const [isOpen, setIsOpen] = useState(false);
  const handleClick = (): void => {
    isOpen ? setIsOpen(false) : setIsOpen(true);
  };

  return (
    <div>
      <Button onClick={handleClick}>Show Popup</Button>
      <Popup isOpen={isOpen} onClosePopup={handleClick}>
        <div className="text-center w-56">
          <div className="mx-auto my-4 w-48">
            <h3 className="text-lg font-black text-gray-800">Confirm Delete</h3>
            <p className="text-sm text-gray-500">Are you sure you want to delete this item?</p>
          </div>
          <div className="flex gap-4 justify-between">
            <Button onClick={handleClick}>Confirm</Button>
            <Button variant="filled" color="danger">
              Cancel
            </Button>
          </div>
        </div>
      </Popup>
    </div>
  );
};

export const Default = Template.bind({});
Default.args = {
  closeButtonContent: 'Close'
};
