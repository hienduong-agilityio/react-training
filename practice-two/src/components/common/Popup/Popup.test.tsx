import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import Popup from './index';

describe('Popup Component', () => {
  it('renders correctly when isOpen is true', () => {
    const onClosePopup = jest.fn();
    const { container, getByText } = render(
      <Popup isOpen={true} onClosePopup={onClosePopup}>
        <div>Popup Content</div>
      </Popup>
    );

    expect(container.firstChild).toHaveClass('visible');
    expect(getByText('Popup Content')).toBeInTheDocument();
  });

  it('renders correctly when isOpen is false', () => {
    const onClosePopup = jest.fn();
    const { container, queryByText } = render(
      <Popup isOpen={false} onClosePopup={onClosePopup}>
        <div>Popup Content</div>
      </Popup>
    );

    expect(container.firstChild).toHaveClass('invisible');
    expect(queryByText('Popup Content')).toBeInTheDocument();
  });

  it('calls onClosePopup when clicking on close button', () => {
    const onClosePopup = jest.fn();
    const { getByText } = render(
      <Popup isOpen={true} onClosePopup={onClosePopup} closeButton={true}>
        <div>Popup Content</div>
      </Popup>
    );

    fireEvent.click(getByText('Closes'));
    expect(onClosePopup).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot', () => {
    const onClosePopup = jest.fn();
    const { asFragment } = render(
      <Popup isOpen={true} onClosePopup={onClosePopup}>
        <div>Popup Content</div>
      </Popup>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
