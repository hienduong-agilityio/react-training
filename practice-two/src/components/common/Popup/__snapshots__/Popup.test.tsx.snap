// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="
          fixed inset-0 flex justify-center items-center transition-colors
          visible bg-black/20
        "
  >
    <div
      class="
            bg-white rounded-xl shadow p-6 transition-all z-50
            scale-100 opacity-100 z-50
          "
    >
      <div>
        Modal Content
      </div>
    </div>
  </section>
</DocumentFragment>
`;
